[android] Use stable manifest ID where applicable

wschurman opened the pull request:
# Why

This PR audits all `RawManifest.getID` callsites and ensures that `scopeKey` (preferred, used in all cases other than expo push registration and auth session) or `originalFullName` are used instead of `id` when possible and applicable. `scopeKey` and `originalFullName` are stable through project transfers so they should be used for client data scoping and legacy API scoping respectively, but should not be used for things like manifest loading or manifest caching at this point.

# How

This PR accomplishes this by performing the following steps.
1. Split `RawManifest.getID` into three new methods and then use the compiler to tell me what needs attention:
    - `getLegacyID` - equivalent to `getID` and used for non-scoped purposes. For example, when storing manifests in the local DB we want to store them by their `id` field still, not the original `id` of the project.
    - `getStableLegacyID` - `originalFullName` with `id` as a fallback. Used for all scoping that previously used `id` for scoping and requires first or third party API calls (amplitude and push notifications for example).
    - `getScopeKey` - `scopeKey` with `getStableLegacyID` as fallback. Used for all scoping on the client that doesn't require external API calls (shared preferences, expo-task-manager scoping, etc...)
2. Replace the `ExperienceId` class (wrapper around the experience id string) with `ExperienceKey`, which is a wrapper around scope key from the manifest. This class is a kotlin data class, so two instances will be equal if their values are equal (same semantics as `ExperienceId` but without the constructor memoization since it seemed unnecessary).
3. Use `ExperienceKey` in as many internal APIs as possible and then only at the leafs where an actual value is needed (to key into shared preferences for example) do we extract the scopeKey.

# Concerns

- At this size, this PR is pretty much guaranteed to have mistakes. This is compounded by the need to manually update 4 versions of the code for every versioned code change. Unfortunately, splitting this up proved to be difficult due to the nature of the change not being able to be done incrementally to retain the safety of the compiler (removing types entirely to find callsites that need attention). 
- Testing this thoroughly is fairly unfeasible as it requires building an app with all the features touched here, transferring the app, running again, creating a new build in the new account, re-running the app, and then pushing an OTA update on the new account for the app. Throughout this four step process we'd need to check all of the following: push notifications, error recovery (we'd need to trigger an error), permissions, sensors, schedulers (scheduled notifications), notification groups, pedometer, facebook, accelerometer, gyroscope, gravity sensor (?), linear acceleration, magnetometer, rotation, tasks (expo-task-manager), and probably some that I've missed.
- Instead of manually testing, here's what I propose for reviewers (I've already run through this, but for code review this is a high leverage review strategy):
    1. Assume compiler is fairly good at figuring out what places needed attention due to method renames (removed `getID` and `ExperienceId` for example).
    2. Check out the branch locally. Use android studio to "Find Usages" of the three methods in `RawManifest` and the corresponding three methods in `ExperienceKey`. Ensure that scopeKey, stableLegacyID, and legacyId are used where expected.

# Test Plan

- Load an experience that has been transferred between accounts, ensure the fields are set to the correct thing in `ExperienceActivity` using a breakpoint. Send a notification to a transferred experience, see that it works.
- The 4 step plan specified in the concerns section.
- Load native component list, use all APIs, see that nothing is broken at first glance, especially the notifications test.
- @tcdavis and I will do a bunch of dogfooding in the future (of all the modules mentioned above) to check all the modules that are scoped in a standalone context, but for now it is tough.

wschurman committed:
[android] Use stable manifest scopeKey instead of id

ide commented:
We should use scopeKey for scoping. It will cause problems if we use originalFullName because in new projects, scopeKey will diverge from full names.

wschurman commented:
> We should use scopeKey for scoping. It will cause problems if we use originalFullName because in new projects, scopeKey will diverge from full names.
> 
> There are two places I know of where we want to use originalFullName right now: auth session (which is all in JS) and notifications (though some parts of notifications should use scopeKey).
> 
> Things like SharedPreferences and Permissions should use scopeKey.

Yeah, I thought about this as I was writing these PRs. I'll see what I can do. The reason I didn't originally is it seemed like we may need to pass scopeKey and originalFullName into scoped modules (`ExpoModuleRegistryAdapter`) but now I remember that those are for Expo Go only so it's fine not to audit them too closely since we're not providing many guarantees for now about transferred projects in Expo Go.

wschurman commented:
Updated this based on review comment, and updated the summary to reflect the new state. I'll hold off on the iOS implementation until this one has settled.

ide commented on android/expoview/src/main/java/host/exp/exponent/experience/ReactNativeActivity.java:
```diff
@@ -610,7 +612,7 @@ public boolean shouldShowRequestPermissionRationale(@NonNull String permission)
     if (requestCode == EXPONENT_PERMISSIONS_REQUEST) {
       @Nullable String name = mManifest.getName();
-      mScopedPermissionsRequester = new ScopedPermissionsRequester(mExperienceId);
+      mScopedPermissionsRequester = new ScopedPermissionsRequester(mStableLegacyExperienceId);
```
Looks like this should use scopeKey

ide commented on android/expoview/src/main/java/host/exp/exponent/experience/ReactNativeActivity.java:
```diff
@@ -635,7 +637,7 @@ public void onRequestPermissionsResult(final int requestCode, final String[] per
   public int checkPermission(final String permission, final int pid, final int uid) {
     int globalResult = super.checkPermission(permission, pid, uid);
-    return mExpoKernelServiceRegistry.getPermissionsKernelService().getPermissions(globalResult, getPackageManager(), permission, mExperienceId);
+    return mExpoKernelServiceRegistry.getPermissionsKernelService().getPermissions(globalResult, getPackageManager(), permission, mStableLegacyExperienceId);
```
Likely should use scope key

ide commented on android/expoview/src/main/java/host/exp/exponent/notifications/ExponentNotificationManager.java:
```diff
@@ -68,7 +68,7 @@ public void maybeCreateNotificationChannelGroup(RawManifest manifest) {
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
       try {
-        String experienceId = manifest.getID();
+        String experienceId = manifest.getStableLegacyID();
```
Skeptical this shouldn't be scope key

ide commented on android/expoview/src/main/java/host/exp/exponent/notifications/NotificationHelper.java:
```diff
@@ -568,7 +568,7 @@ public static void scheduleLocalNotification(
 
     try {
-      experienceId = manifest.getID();
+      experienceId = manifest.getStableLegacyID();
```
Likely should use scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/ExponentPackage.java:
```diff
@@ -189,7 +189,7 @@ public static ExponentPackage kernelExponentPackage(Context context, RawManifest
       try {
-        ExperienceId experienceId = ExperienceId.create(mManifest.getID());
+        ExperienceId experienceId = ExperienceId.create(mManifest.getStableLegacyID());
         ScopedContext scopedContext = new ScopedContext(reactContext, experienceId.getUrlEncoded());
```
ScopedContext almost certainly should use scopeKey

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/PedometerModule.java:
```diff
@@ -218,7 +218,7 @@ public void onHostDestroy() {
     try {
       ExperienceActivity activity = (ExperienceActivity) getCurrentActivity();
-      return activity != null ? activity.getExperienceId() : null;
+      return activity != null ? activity.getLegacyExperienceId() : null;
```
This should use scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -96,7 +96,7 @@ public void deleteCategoryAsync(final String categoryIdParam, final Promise prom
     if (!Constants.isStandaloneApp()) {
       try {
-        String experienceId = mManifest.getID();
+        String experienceId = mManifest.getStableLegacyID();
```
Category-related things should use scope key (we are going to need to send down the scope key in notification payloads)

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -144,7 +144,7 @@ public void getExponentPushTokenAsync(final Promise promise) {
 
     try {
-      String experienceId = mManifest.getID();
+      String experienceId = mManifest.getStableLegacyID();
```
This is one of the very few places where stable legacy ID seems right for now

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -198,7 +198,7 @@ public void deleteChannel(String channelId, final Promise promise) {
 
     try {
-      experienceId = mManifest.getID();
+      experienceId = mManifest.getStableLegacyID();
```
scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -233,7 +233,7 @@ public void presentLocalNotificationWithChannel(final ReadableMap data, final Re
 
     try {
-      experienceId = mManifest.getID();
+      experienceId = mManifest.getStableLegacyID();
```
scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -281,7 +281,7 @@ public void scheduleLocalNotificationWithChannel(final ReadableMap data, final R
       String experienceId;
       try {
-        experienceId = mManifest.getID();
+        experienceId = mManifest.getStableLegacyID();
```
scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -328,7 +328,7 @@ public void dismissNotification(final int notificationId, final Promise promise)
       ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
       manager.cancel(
-        mManifest.getID(),
+        mManifest.getStableLegacyID(),
```
likely scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -341,7 +341,7 @@ public void dismissNotification(final int notificationId, final Promise promise)
     try {
       ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
-      manager.cancelAll(mManifest.getID());
+      manager.cancelAll(mManifest.getStableLegacyID());
```
likely scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -352,7 +352,7 @@ public void dismissAllNotifications(final Promise promise) {
     try {
       ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
-      manager.cancelScheduled(mManifest.getID(), notificationId);
+      manager.cancelScheduled(mManifest.getStableLegacyID(), notificationId);
```
likely scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -375,9 +375,9 @@ public void cancelScheduledNotificationWithStringIdAsync(final String id, final
     try {
       ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
-      manager.cancelAllScheduled(mManifest.getID());
+      manager.cancelAllScheduled(mManifest.getStableLegacyID());
```
likely scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -375,9 +375,9 @@ public void cancelScheduledNotificationWithStringIdAsync(final String id, final
+      manager.cancelAllScheduled(mManifest.getStableLegacyID());
 
-      String experienceId = mManifest.getID();
+      String experienceId = mManifest.getStableLegacyID();
```
scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -402,7 +402,7 @@ public void scheduleNotificationWithTimer(final ReadableMap data, final Readable
 
     try {
-      experienceId = mManifest.getID();
+      experienceId = mManifest.getStableLegacyID();
```
scope key, but may require some changes to IntervalSchedulerModel

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -445,7 +445,7 @@ public void scheduleNotificationWithCalendar(final ReadableMap data, final Reada
 
     try {
-      experienceId = mManifest.getID();
+      experienceId = mManifest.getStableLegacyID();
```
scope key

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/internal/ExponentAsyncStorageModule.java:
```diff
@@ -29,7 +29,7 @@ public ExponentAsyncStorageModule(ReactApplicationContext reactContext, RawManif
 
     try {
-      String experienceId = manifest.getID();
+      String experienceId = manifest.getStableLegacyID();
```
scope key

ide reviewed (changes requested):
Most of the places that use Stable Legacy ID should use Scope Key instead. The only places I know of so far that use Stable Legacy ID are:

- Registering for push notifications, since the server expects a full name and uses it for credential lookups.
- AuthSession URLs, since the proxy server expects a full name.

The common pattern is that both the push notification server and the AuthSession server are external services that explicitly expect a full name instead of an arbitrary string. (In fact, both of these services want to identify the projects, so EAS Project ID will be the appropriate value to use for new apps.)

There may be other APIs that need Stable Legacy ID, but we should first see if Scope Key works. Scope Key is used to differentiate data belonging to different projects. Stable Legacy ID is can do that, too, but should be used only when Scope Key cannot be made to work -- very rare, I think.

---

The ExperienceKey class looks OK. Perhaps eventually it will be replaced by passing around just Scope Key. The few APIs that need EAS Project ID perhaps should access it another way (ex: `manifest.getExtraField("eas", "projectId")`) so that EAS-proprietary fields stay separated out.

---

The change from appId to appScopeKey looks OK but in some places, just `scopeKey` instead of `appScopeKey` might line up better with how we call the field just "scope key" in most places. I don't think this is a big deal though.

esamelson commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/internal/ExponentAsyncStorageModule.java:
```diff
@@ -15,22 +15,22 @@
-    String experienceIdEncoded = URLEncoder.encode(experienceId, "UTF-8");
-    return "RKStorage-scoped-experience-" + experienceIdEncoded;
+  public static String experienceKeyToDatabaseName(ExperienceKey experienceKey) throws UnsupportedEncodingException {
+    return "RKStorage-scoped-experience-" + experienceKey.getUrlEncodedScopeKey();
```
Are we confident this isn't used in standalone (turtle v1) apps? If it is, we need to use the stable legacy ID here.

esamelson commented on android/expoview/src/main/java/host/exp/exponent/utils/ScopedContext.java:
```diff
@@ -57,8 +59,11 @@
+  public ScopedContext(final Context context, final ExperienceKey experienceKey) throws UnsupportedEncodingException {
     super(context);
+
+    String scope = experienceKey.getUrlEncodedScopeKey();
```
This seems like a particularly dangerous change given how many modules access shared preferences through this class. We may need to use the legacy stable ID here as well if any of those modules access this class in standalone apps (which I'm not seeing any reason they wouldn't, but I might be blind).

ide replied:
We almost never want to use the stable legacy ID field. The scope key will be the correct value here.

ide replied:
We almost never want to use the stable legacy ID field. The scope key will be the correct value here.

esamelson replied:
The actual value of the scope key is different from the stable legacy ID, even for apps that existed before any of these changes, correct? I'm just saying that regardless of which value is ideologically correct to use here, I think it's quite possible that changing this value will break things in standalone apps, for modules that access SharedPreferences through this ScopedContext (which there are a number of).

ide replied:
No, the scope key for old projects is equal to their original full name i.e. the stable legacy ID.

The scope key for new projects will be a separate, dedicated value.

esamelson replied:
Oh 🤦 well in that case I think basically all my comments can be disregarded. Will edit as such. Sorry for the confusion.

esamelson reviewed (commented):
Generally this looks ok to me, though would emphasize what you have already said about dogfooding, especially in a standalone context.

ed: basically all my comments were based on an incorrect assumption about the scope key value. I think this looks fine.

ide commented on android/expoview/src/main/java/host/exp/exponent/notifications/NotificationConstants.java:
```diff
@@ -5,7 +5,7 @@
   public static final int MAX_COLLAPSED_NOTIFICATIONS = 5;
   public static final String NOTIFICATION_MESSAGE_KEY = "message"; // deprecated
-  public static final String NOTIFICATION_EXPERIENCE_ID_KEY = "experienceId";
+  public static final String NOTIFICATION_EXPERIENCE_ID_KEY = "experienceId"; // scopeKey
```
Despite the awkward "key key", it seems more correct to rename this field:
```suggestion
  public static final String NOTIFICATION_SCOPE_KEY_KEY = "experienceId"; // scopeKey
```

ide commented on android/expoview/src/main/java/host/exp/exponent/notifications/NotificationHelper.java:
```diff
@@ -372,9 +372,9 @@ public static void showNotification(
           final Listener listener) {
     final ExponentNotificationManager manager = new ExponentNotificationManager(context);
-    final String experienceId = (String) details.get("experienceId");
+    final String experienceScopeKey = (String) details.get("experienceId");
```
Looking at other call sites, I believe it's correct to use `NotificationConstants` to key into the details HashMap:

```suggestion
    final String experienceScopeKey = (String) details.get(NotificationConstants. NOTIFICATION_EXPERIENCE_ID_KEY);
```

ide commented on android/expoview/src/main/java/host/exp/exponent/notifications/NotificationHelper.java:
```diff
@@ -575,19 +575,8 @@ public static void scheduleLocalNotification(
-      listener.onFailure(new Exception("Requires Experience Id"));
-      return;
-    }
+    details.put("experienceId", experienceKey.getScopeKey());
```
Use `NotificationConstants` for the map keys? (We should verify whether the receiving side that reads from `details` uses `NotificationConstants` so we maintain symmetry.)

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java:
```diff
@@ -229,13 +210,7 @@ public void presentLocalNotificationWithChannel(final ReadableMap data, final Re
-      promise.reject("E_FAILED_PRESENTING_NOTIFICATION", "Requires Experience ID");
-      return;
-    }
+    details.put("experienceId", mExperienceKey.getScopeKey());
```
`NotificationConstants` keys? (Same throughout this class)

ide commented on android/expoview/src/main/java/host/exp/exponent/notifications/PushNotificationHelper.java:
```diff
@@ -164,7 +167,7 @@ public void onLoadBitmap(final Bitmap bitmap) {
         // Create notification object
         boolean isMultiple = mode == Mode.COLLAPSE && unreadNotifications.length() > 1;
-        final ReceivedNotificationEvent notificationEvent = new ReceivedNotificationEvent(experienceId, body, notificationId, isMultiple, true);
+        final ReceivedNotificationEvent notificationEvent = new ReceivedNotificationEvent(experienceKey.getStableLegacyId(), body, notificationId, isMultiple, true);
```
`ReceivedNotificationEvent` now takes a scope key:
```suggestion
        final ReceivedNotificationEvent notificationEvent = new ReceivedNotificationEvent(experienceKey.getScopeKey(), body, notificationId, isMultiple, true);
```

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFirebaseCoreService.java:
```diff
@@ -86,14 +85,14 @@ public ScopedFirebaseCoreService(Context context, RawManifest manifest, Experien
-  private static String getEncodedExperienceId(ExperienceId experienceId) {
+  private static String getEncodedExperienceId(ExperienceKey experienceKey) {
     try {
-      String encodedUrl = experienceId.getUrlEncoded();
+      String encodedUrl = experienceKey.getUrlEncodedStableLegacyId();
       byte[] data = encodedUrl.getBytes("UTF-8");
       String base64 = Base64.encodeToString(data, Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP);
       return base64;
     } catch (UnsupportedEncodingException e) {
-      return Integer.toString(experienceId.hashCode());
+      return Integer.toString(experienceKey.getStableLegacyId().hashCode());
     }
   }
```
I don't want to hold up this whole PR but I think we should have a focused discussion about what ID is correct to use here, assuming we phase out full names for new projects.

ide reviewed (changes requested):
Thinking things through, I really like the ExperienceKey class as a stepping stone to make the type system differentiate scope keys from experience IDs in Java without nominal type aliases (`type ScopeKey = string`).

wschurman commented:
I'm changing notification stuff a bit, mainly changing `ExperienceDB` to support project IDs now that https://github.com/expo/universe/pull/7665 has landed in prod. Will update this PR when ready.

ide commented:
> mainly changing ExperienceDB to support project IDs

Let's discuss this today. I think a big question is: do we want projects to need an EAS project ID in order to work with Expo Go? Or do they just need a scope key?

ide commented on android/expoview/src/main/java/host/exp/exponent/kernel/services/ErrorRecoveryManager.java:
```diff
@@ -11,38 +11,38 @@
   private static final long AUTO_RELOAD_BUFFER_BASE_MS = 5 * 1000;
 
-  private static Map<ExperienceId, ErrorRecoveryManager> sExperienceIdToManager = new HashMap<>();
+  private static Map<ExperienceKey, ErrorRecoveryManager> sExperienceKeyToManager = new HashMap<>();
```
I think this should be a map from scope key to ErrorRecoveryManager because conceptually speaking, if someone were to rename/transfer a project, the scope key would remain constant and be the value we want to use as the key. In other words, we want to scope ErrorRecoveryManagers so scope key is best.

ide commented on android/expoview/src/main/java/host/exp/exponent/kernel/services/sensors/SubscribableSensorKernelService.java:
```diff
@@ -12,26 +12,26 @@
   protected static int DEFAULT_UPDATE_INTERVAL = 100;
 
-  private Map<ExperienceId, Integer> mExperienceIdListenersCountMap = new HashMap<>();
+  private Map<ExperienceKey, Integer> mExperienceKeyListenersCountMap = new HashMap<>();
```
While this works because the device sensor subscriptions are scoped to the lifetime of the process, scope key is conceptually more appropriate here since a project's subscriptions shouldn't change if the project is renamed or transferred.

(This is a philosophical point right now but if someone were to copy it to a module that retains state across launches, using the ExperienceKey to scope would not be transfer-safe.)

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/PedometerModule.java:
```diff
@@ -44,13 +44,14 @@
   private @Nullable OnDataPointListener mListener;
   private int mWatchTotalSteps = 0;
-  private static Map<String, GoogleApiClient> sInstanceMap = new HashMap<>();
+  private static Map<ExperienceKey, GoogleApiClient> sInstanceMap = new HashMap<>();
```
Another candidate to be keyed by scope key instead of ExperienceKey

ide commented on android/expoview/src/main/java/host/exp/exponent/experience/ExperienceActivity.java:
```diff
@@ -579,9 +578,8 @@ public void setManifest(String manifestUrl, final RawManifest manifest, final St
-      String id;
       try {
-        id = Exponent.getInstance().encodeExperienceId(mExperienceIdString);
+        Exponent.encodeExperienceId(mExperienceKey.getLegacyId());
```
Do we need this? Was the intent to make sure the scope key is URL-encodable? (For example, ScopedContext's constructor URL-encodes the scope key.)

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFirebaseCoreService.java:
```diff
@@ -86,14 +85,14 @@ public ScopedFirebaseCoreService(Context context, RawManifest manifest, Experien
+  private static String getEncodedExperienceId(ExperienceKey experienceKey) {
     try {
-      String encodedUrl = experienceId.getUrlEncoded();
+      String encodedUrl = experienceKey.getUrlEncodedStableLegacyId();
```
Thinking more about this, this needs to be the scope key because modern manifests won't necessarily have a legacy ID (full name).

We should keep in mind that we want our scope keys to be nicely URL-encodable (ideally URL-safe to begin with).

wschurman commented:
I can simplify this a bit further.

ide commented on android/expoview/src/main/java/host/exp/exponent/experience/HomeActivity.java:
```diff
@@ -69,7 +69,7 @@ protected void onCreate(Bundle savedInstanceState) {
       mExperienceKey = ExperienceKey.fromRawManifest(mManifest);
     } catch (JSONException e) {
-      mExperienceKey = new ExperienceKey("", "", "");
+      mExperienceKey = new ExperienceKey( "", "");
```
```suggestion
      mExperienceKey = new ExperienceKey("", "");
```

ide reviewed (approved):
This looks good to me! I recognize how much behavior across codebases and across time -- existing apps -- needed to be researched and taken into consideration to get here. What a journey so far.

wschurman commented:
Going to rebase (we've since cut a new version so I expect this to take a while), self-audit once more, and depending on how that goes will update the test plan, sanity check push notifications since that had the most significant changes, and then land or request re-review if necessary.

wschurman committed:
Fix service error

wschurman commented:
All these packages are broken in master. CI in master only runs for changed files though, and it seems that these failures aren't due to files changed in the packages, but due to dependencies I guess.

Also the client build is broken in master as well.

Planning to land this as-is late Monday unless there are any objections so that SDK 42 can be cut mid-day Monday at the commit before this one. This is desirable since although this commit will make cherry-picking a pain due to merge conflicts, it also gives us the most time in master to iron out any issues and implement the equivalent for iOS in the same SDK release (the next one, 43).

wschurman committed:
Changelogs

dqii commented:
https://github.com/expo/expo/issues/15158#issuecomment-1017726258
I think this might have caused some breaks in FCM push notifications?
